extensions-core/multi-stage-query/src/test/java/org/apache/druid/msq/kernel/controller/WorkerInputsTest.java

```diff
@@ -19,14 +19,369 @@
+    );
+
+    Assert.assertEquals(
+        ImmutableMap.<Integer, List<InputSlice>>builder()
+                    .put(0, Collections.singletonList(NilInputSlice.INSTANCE))
+                    .build(),
+        inputs.assignmentsMap()
+    );
+  }
+
+  @Test
+  public void test_auto_zeroInputSlices_broadcast_fourWorkers()
+  {
+    final StageDefinition stageDef =
+        StageDefinition.builder(0)
+                       .inputs(new TestInputSpec())
+                       .broadcastInputs(IntSet.of(0))
+                       .maxWorkerCount(4)
+                       .processorFactory(new OffsetLimitFrameProcessorFactory(0, 0L))
+                       .build(QUERY_ID);
+
+    final WorkerInputs inputs = WorkerInputs.create(
+        stageDef,
+        Int2IntMaps.EMPTY_MAP,
+        new TestInputSpecSlicer(true),
+        WorkerAssignmentStrategy.AUTO
+    );
+
+    Assert.assertEquals(
+        ImmutableMap.<Integer, List<InputSlice>>builder()
+                    .put(0, Collections.singletonList(new TestInputSlice()))
+                    .build(),
+        inputs.assignmentsMap()
+    );
+  }
+
+  @Test
+  public void test_auto_threeInputs_fourWorkers()
+  {
+    final StageDefinition stageDef =
+        StageDefinition.builder(0)
+                       .inputs(new TestInputSpec(1, 2, 3))
+                       .maxWorkerCount(4)
+                       .processorFactory(new OffsetLimitFrameProcessorFactory(0, 0L))
+                       .build(QUERY_ID);
+
+    final WorkerInputs inputs = WorkerInputs.create(
+        stageDef,
+        Int2IntMaps.EMPTY_MAP,
+        new TestInputSpecSlicer(true),
+        WorkerAssignmentStrategy.AUTO
+    );
+
+    Assert.assertEquals(
+        ImmutableMap.<Integer, List<InputSlice>>builder()
+                    .put(0, Collections.singletonList(new TestInputSlice(1, 2, 3)))
+                    .build(),
+        inputs.assignmentsMap()
+    );
+  }
+
+  @Test
+  public void test_auto_threeBigInputs_fourWorkers()
+  {
+    final StageDefinition stageDef =
+        StageDefinition.builder(0)
+                       .inputs(new TestInputSpec(4_000_000_000L, 4_000_000_001L, 4_000_000_002L))
+                       .maxWorkerCount(4)
+                       .processorFactory(new OffsetLimitFrameProcessorFactory(0, 0L))
+                       .build(QUERY_ID);
+
+    final WorkerInputs inputs = WorkerInputs.create(
+        stageDef,
+        Int2IntMaps.EMPTY_MAP,
+        new TestInputSpecSlicer(true),
+        WorkerAssignmentStrategy.AUTO
+    );
+
+    Assert.assertEquals(
+        ImmutableMap.<Integer, List<InputSlice>>builder()
+                    .put(0, Collections.singletonList(new TestInputSlice(4_000_000_000L, 4_000_000_001L)))
+                    .put(1, Collections.singletonList(new TestInputSlice(4_000_000_002L)))
+                    .build(),
+        inputs.assignmentsMap()
+    );
+  }
+
+  @Test
+  public void test_auto_threeBigInputs_oneWorker()
```

Comment:
Do these tests cover the limits? That is, more splits than the limit? More capacity than available?

Do the tests cover skew: Many small files and one huge one?

Reply:
What do you mean by "more splits than the limit" and "more capacity than available"?

I'll add a new test for many small files & one big one, it will be called `test_auto_tenSmallAndOneBigInput_threeWorker`.

Reply:
Those two comments were trying to work out the limits of the split algorithm. What happens if there is more data than fits into the available number of splits, each filled to the maximum size? Or, is there a maximum size per split? If there is more data, do we keep adding more to the existing splits round-robin?

Reply:
Ah, I see. The intent is that if there is more data, then we keep adding more to the existing splits, lightest-split-first. This test (`test_auto_threeBigInputs_oneWorker`) covers the behavior.